hilti/toolchain/src/compiler/jit.cc

```diff
@@ -18,6 +23,37 @@
 
 using namespace hilti;
 
+namespace {
+
+std::string readFile(const hilti::rt::filesystem::path& path) {
+    std::ifstream ifs(path);
+    return {(std::istreambuf_iterator<char>(ifs)), (std::istreambuf_iterator<char>())};
+}
+
+hilti::rt::filesystem::path save(const CxxCode& code, const hilti::rt::filesystem::path& id, std::size_t hash) {
+    const std::size_t cc_hash = code.hash();
+
+    const auto cc = hilti::rt::filesystem::temp_directory_path() /
+                    util::fmt("%s_%" PRIx64 "-%" PRIx64 ".cc", id.stem().c_str(), hash, cc_hash);
+
+    HILTI_DEBUG(logging::debug::Jit, util::fmt("writing temporary code for %s to %s", id, cc.filename().native()));
+    const auto file = std::fopen(cc.c_str(), "w");
+    if ( ! file )
+        rt::fatalError(util::fmt("could not open file '%'s for writing", cc));
+
+    if ( const auto& content = code.code() ) {
+        if ( std::fputs(content->c_str(), file) == EOF )
+            rt::fatalError(util::fmt("could not write to temporary file '%s'", cc));
+    }
+
+    if ( std::fflush(file) == EOF )
+        rt::fatalError(util::fmt("could not finalize write to temporary file '%s': %s", cc, std::strerror(errno)));
+
+    return cc;
```

Comment:
We still could race on the output file if multiple jobs with identical flags from the same directory try to write to the same file. Let's instead write to some temp file first and then move it to the final location atomically. This probably needs additional gc for the case where we fail before moving the temp file over.